### PR TITLE
OSM Map into next branch

### PR DESCRIPTION
**Pull request by adamw523**
Pulled in next branch off blackberry and merged OSM addition to it.

Hoping this is what Ken is requesting. Let me know if it's not

### CODE REVIEW

**Comment by brentlintner** [comment]
Thanks for being patient and also rebasing @adamw523. @kwallis are we good to move forward?

**Comment by kwallis** [comment]
Hi Adam.  As Next represents our next release, which we are hoping to finalize soon, I was hoping you could have this in an osm branch off of next, not in next directly.  That will allow us to complete the pull request, but not have it go directly into next.  Please let me know if I am not making sense and I can walk you through it.

**Comment by brentlintner** [comment]
Hey I wonder if anyone else can reproduce this (web and extension). I found a weird issue _only_ when you initialize Ripple-UI with the Geo Location pane closed. If you open it up, the Map image is missing, and there is an exception in OpenLayers:

```
Uncaught TypeError: Cannot read property 'w' of null
    OpenLayers.Map.OpenLayers.Class.calculateBoundsripple.js:22836
    OpenLayers.Layer.OpenLayers.Class.getExtentripple.js:23127
    OpenLayers.Map.OpenLayers.Class.getExtentripple.js:22870
    OpenLayers.Map.OpenLayers.Class.moveToripple.js:22858
    OpenLayers.Map.OpenLayers.Class.setCenterripple.js:22840
    _updateGpsMapripple.js:40371
    module.exports.initialize.event.trigger.latituderipple.js:40602
    module.exports.on
```

Perhaps the display: block CSS state might be causing issues. Going to keep looking into it, thought I would post as well. (cc @gtanner @ndaversa)

**Comment by adamw523** [comment]
Ken, It's probably quickest if you walk me through the git magic of
setting that up.. or at least provide a list of commands that will
accomplish that.

On Tue, Sep 27, 2011 at 5:24 PM, Ken Wallis
reply@reply.github.com
wrote:

**Comment by kwallis** [comment]
Hi Adam.  I went ahead and created the osm branch.  It is a branch off of next, and would be the best place for you to pull this in.  Sorry for the back and forth and overhead.  I hope this is the last rebase etc. we will need to worry about.  Really excited to see this pulled in!

**Comment by brentlintner** [comment]
Hey Ken, I think I get what you mean, but I think to do that, one of us needs to pull this branch locally and technically push a copy of it into blackberry/Ripple-UI, or create a branch off the same commit SHA (ID) that osm_next is, push it and then ask Adam to update the pull request to point to that instead of next, and merge it in, or create a patch, or probably some other way that I don't know heh.

Either way, it would not be difficult to do (just tedious IMO), but I am wondering what difference it makes to do so? I.e. VS Leaving this open, and pulling it into next (or having to rebase again because if it's left too long) when we are able to. That way, IMO we don't lose track of code waiting to get into a release. I suppose a parallel example would be a feature branch off of next in the blackberry-webworks fork needing to be pulled into blackberry/Ripple-UI.

Thoughts?

**Comment by brentlintner** [comment]
@kwallis posted at the same time, didn't see your reply lol. My question still stands though. :-)

**Comment by kwallis** [comment]
The feature branch model we are using in blackberry-webworks is what I was
trying to describe.  I created the osm feature branch off of next now.

But I do feel we will need to consider the best ways to handle pull
requests like these.  The committers on these projects may have to
undertake some additional overhead and do some manual magic behind the
scenes, so we don't have to make these types of requests back to the
contributors.  I guess we are working through the process and can still
refine it. I am not sure leaving Pull requests open for extended periods
is the best thing eitherŠ  Open to ideas.

Thanks Adam for being so patient and helping us with this. ;)

## 

Ken Wallis
Manager ­ BlackBerry WebWorks

Research In Motion
(905) 629-4746 x14369

-----Original Message-----
From: Brent Lintner
reply@reply.github.com
Date: Tue, 27 Sep 2011 18:03:11 -0400
To: Ken Wallis kwallis@rim.com
Subject: Re: [Ripple-UI] OSM Map into next branch (#163)

> Hey Ken, I think I get what you mean, but I think to do that, one of us
> needs to pull this branch locally and technically push a copy of it into
> blackberry/Ripple-UI, or create a branch off the same commit SHA (ID)
> that osm_next is, push it and then ask Adam to update the pull request to
> point to that instead of next, and merge it in, or create a patch, or
> probably some other way that I don't know heh.
> 
> Either way, it would not be difficult to do (just tedious IMO), but I am
> wondering what difference it makes to do so? I.e. vs Leaving this open,
> and pulling it into next (or having to rebase again because if it's left
> too long) when we are able to. That way, IMO we don't lose track of code
> waiting to get into a release. I suppose a parallel example would be a
> feature branch off of next in the blackberry-webworks fork needing to be
> pulled into blackberry/Ripple-UI.
> 
> Thoughts?
> 
> ## 
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/blackberry/Ripple-UI/pull/163#issuecomment-2217116

---

This transmission (including any attachments) may contain confidential information, privileged material (including material protected by the solicitor-client or other applicable privileges), or constitute non-public information. Any use of this information by anyone other than the intended recipient is prohibited. If you have received this transmission in error, please immediately reply to the sender and delete this information from your system. Use, dissemination, distribution, or reproduction of this transmission by unintended recipients is not authorized and may be unlawful.

**Comment by adamw523** [comment]
A downside of a feature branch on bb/Ripple-UI is that it will quickly
fall behind next.

Could I just branch of next in my own feature branch and then request
a pull into a different branch on bb/Ripple-UI.. Possibly an
"external-pulls" branch?

That way "next" is always up to date, and you can update
"external-pulls" with "next" only when you're working on pull request.

This sounds like it should be doable...

On Tue, Sep 27, 2011 at 6:09 PM, Ken Wallis
reply@reply.github.com
wrote:

**Comment by brentlintner** [comment]
Ah ok cool, thanks for clarifying that.

I agree that the process could be refined or changed though. I still strongly suggest we should keep it open instead of duplicating effort by merging it in and keeping track of it outside our main issue tracker. :-s

I.e., IMHO this will fall apart when we have a lot or even a bunch of pull requests though. Having a pull request open is essentially a placeholder "Issue" (that is auto created on GitHub as well) for a branch that is waiting to be pulled in, whether in the main repo, or a fork. Merging it in closes the Issue created by it (and the Pull Request itself), which seems counter intuitive to closing an Issue when it's still in a feature branch and not being shipped (i.e. it's in next).

This also means we have duplicate branches in the central fork that sit around until they are rebased/merged into next. :-s

Perhaps we can talk about this tomorrow or sometime? :-) I can't help but see the possibility for a lot of disconnects/issues with certain aspects of our current strategy.

I'm also open to thoughts from the community/everyone as well! :-)

**Comment by brentlintner** [comment]
oops, @adamw523, I just saw your comment after I posted mine.

**Comment by adamw523** [comment]
I've wrapped the call to setCenter on the map object with a " jQuery(mapContainer).is(':visible') " call, so the error reported earlier isnt' thrown when the geo panel is hidden on load

but, once the geo panel is shown, the map isn't  centered on anything...

is there a way to act on an event when the geo panel is shown? other than doing:
event.on("ApplicationState" ...

**Comment by brentlintner** [comment]
Alas, it would be sweet to get OpenLayers to play nice with a hidden element (or at least I think that is what is happening)!

Anyways, I think you could figure out when the panel is shown by something similar to here: https://github.com/blackberry/Ripple-UI/blob/next/lib/ripple/ui.js#L204

I.e. An idea (IMO) would be to listen to that event, and check this.id === "gps-container", and also if it has a "ui-box-open" class (see just a bit below the line I mentioned), or another way even.

Thoughts?

**Comment by adamw523** [comment]
"ApplicationState" sounded so generic that I didn't want to watch on it.. but I guess it's not used as much as I expected. So I've pushed up a fix to my osm branch, watching on "ApplicationState" as you suggested.

at: https://github.com/adamw523/Ripple-UI/commit/e8ad2c84ed2a68636a84394217e1f078dbbe6fc4#L0R222
I've added [pNode] to the event.trigger call. passing "this" might be a better generic way, if we don't mind doing the lookups for the <section> again. 

Let me know if there's any feedback.  

Also, let me know how we're merging... If things have been decided.

**Comment by brentlintner** [comment]
Sweet, that should do the trick! However, I think you pushed to osm and not osm_next (as the diff you mentioned did not show up in this request). To test it, I ended up copying over the diffs from that commit into osm_next, but I think I may have missed something, since I still get that exception when it's initially closed and it does not initialize. (?)

```
Uncaught TypeError: Cannot read property 'w' of null
```

Also, just to clarify, I believe I was a bit vague when I conveyed my idea, as I was suggesting to listen to the .click event (on that line I mentioned) instead of ApplicationState (apologies if so). Regardless, seems like it would work just as well though. :-) The only thing I would (personally) nitpick would be passing pNode to ApplicationState (i.e. you could invoke something like [YAGNI](http://en.wikipedia.org/wiki/You_ain't_gonna_need_it) on it), but whichever really (just my feedback).

**Comment by brentlintner** [comment]
Oh, and regarding merging, this can stay open, and when applicable, we can merge it in. Hopefully that is sooner than later. :-)

**Comment by adamw523** [comment]
I've pushed to osm_next now .. let me know if you're still getting the error as things are working fine for me.

I didn't want to attach to a .click event because that felt like that would be coupling to UI elements too much in the goeView lib. I'm imagining that other events than just clicking on that bar would cause the goeView to appear. 

Passing pNode with ApplicationState does feel a little wrong, but it was the best I could come up with at the time

**Comment by brentlintner** [comment]
@kwallis do you know about any updates on this? (thanks!) I assume the contributor agreement is done. I was hoping to merge this awesomeness in (i.e. even this weekend) if possible. Seems solid to me (and way better than the current maps!).

**Comment by kwallis** [comment]
Sorry I didn't see this before the weekend. ;)  CLA is done.  Once we have
completed our latest internal build and got the current next branch pulled
in internally, let's go ahead and merge this.  @ndaversa or @gtanner can
advise when the internal build is complete.

## Thanks!

Ken Wallis
Manager ­ BlackBerry WebWorks

Research In Motion
(905) 629-4746 x14369

-----Original Message-----
From: Brent Lintner
<reply+i-1748952-63237d265b1b39f0f9c83cc04d768fb8e724f4f6-498318@reply.gith
ub.com>
Date: Fri, 11 Nov 2011 22:00:33 -0500
To: Ken Wallis kwallis@rim.com
Subject: Re: [Ripple-UI] OSM Map into next branch (#163)

> @kwallis do you know about any updates on this? (thanks!) I assume the
> contributor agreement is done. I was hoping to merge this awesomeness in
> (i.e. even this weekend) if possible. Seems solid to me (and way better
> than the current maps!).
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/blackberry/Ripple-UI/pull/163#issuecomment-2715795

---

This transmission (including any attachments) may contain confidential information, privileged material (including material protected by the solicitor-client or other applicable privileges), or constitute non-public information. Any use of this information by anyone other than the intended recipient is prohibited. If you have received this transmission in error, please immediately reply to the sender and delete this information from your system. Use, dissemination, distribution, or reproduction of this transmission by unintended recipients is not authorized and may be unlawful.

File: lib/ripple/ui/plugins/geoView.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 /*
+    z-index: 800;
```
Comment (brentlintner) [inline lib/ripple/ui/plugins/geoView.js]: Just saw this btw. ? :-)

**Comment by brentlintner** [comment]
@adamw523 Apologies, unfortunately no one thought to verify Open Layers can be added to Ripple-UI as a thirdparty library until now. Hopefully we can get this in a release asap. but until then this still needs to go through some internal processes, and I am not too sure how long it will take. :-s Appreciate your patience.

**Comment by kwallis** [comment]
We're pushing to have this done as soon as possible.  I really want to
have this integrated as soon as we can.  Thanks!

## 

Ken Wallis
Manager ­ BlackBerry WebWorks

Research In Motion
(905) 629-4746 x14369

-----Original Message-----
From: Brent Lintner
<reply+i-1748952-63237d265b1b39f0f9c83cc04d768fb8e724f4f6-498318@reply.gith
ub.com>
Date: Tue, 22 Nov 2011 11:27:37 -0500
To: Ken Wallis kwallis@rim.com
Subject: Re: [Ripple-UI] OSM Map into next branch (#163)

> @adamw523 Apologies, unfortunately no one thought to verify Open Layers
> can be added to Ripple-UI as a thirdparty library until now. Hopefully we
> can get this in a release asap. but until then this still needs to go
> through some internal processes, and I am not too sure how long it will
> take. :-s Appreciate your patience.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/blackberry/Ripple-UI/pull/163#issuecomment-2836406

---

This transmission (including any attachments) may contain confidential information, privileged material (including material protected by the solicitor-client or other applicable privileges), or constitute non-public information. Any use of this information by anyone other than the intended recipient is prohibited. If you have received this transmission in error, please immediately reply to the sender and delete this information from your system. Use, dissemination, distribution, or reproduction of this transmission by unintended recipients is not authorized and may be unlawful.

File: lib/ripple/ui/plugins/geoView.js (hunk 1 above)
Comment (adamw523) [inline lib/ripple/ui/plugins/geoView.js]: :-/ weird :)... removing..

**Comment by dansilivestru** [comment]
Hi, Open Layers has been approved as an open source component we can use in Ripple. @kwallis unless you have any objections, I would like to merge this pull request in today.

**Pull request merged**